File: app_script/src/repositories/_Repository.ts
import Spreadsheet = GoogleAppsScript.Spreadsheet.Spreadsheet;
import Sheet = GoogleAppsScript.Spreadsheet.Sheet;

export abstract class Repository<T> {
  abstract toValue?(object: any[]): T | null;

  protected constructor(public ss: Spreadsheet,
                        public name: string,
                        private firstContentIndex: number = 1) {
  }

  getSheet(): Sheet {
    return this.ss.getSheetByName(this.name);
  }

  getAll(): T[] {
    return this.findAll(() => true);
  }

  findAll(predicateFn: (value: any[], index: number, array: any[][]) => boolean): T[] {
    const sheet = this.getSheet();
    const self = this;

    function filterFct(value: any[], index: number, array: any[][]) {
      return index >= self.firstContentIndex &&
        predicateFn(value, index, array)
    }

    const values = sheet.getDataRange().getValues();
    return values.filter(filterFct).map(this.toValue).filter((value) => value);
  }
}

File: app_script/lib/src/data/repository.dart
import 'package:collection/collection.dart';
import 'package:google_apps_script/google_apps_script.dart';

abstract class Repository<R> {
  R? fromRow(List<dynamic> row);

  List<dynamic> toRow(R value);

  const Repository({
    required this.ss,
    required this.name,
    this.firstContentIndex = 1,
  });

  final Spreadsheet ss;
  final String name;
  final int firstContentIndex;

  Sheet get sheet {
    return ss.getSheetByName(name);
  }

  Iterable<R> getAll() {
    return findAll((_) => true);
  }

  Iterable<R> findAll(bool Function(List<dynamic> value) predicate) {
    return sheet //
        .getDataRange()
        .getValues()
        .whereIndexed((index, element) => index >= firstContentIndex && predicate(element))
        .map(fromRow)
        .whereType<R>();
  }
}
